File: 59. Spiral matrix II/solution.py
from enum import Enum
class Dir(Enum):
    Up = 0
    Down = 1
    Left = 2
    Right = 3
class Solution:
    def generateMatrix(self, n: int) -> List[List[int]]:
        res = [[0] * n for _ in range(n)]
        up,down,left,right = 0,n,-1,n
        direction = Dir.Right
        i = 0
        j = 0
        x = 1
        while x <= n**2:
            match direction:
                case Dir.Right:
                    while j < right:
                        res[i][j] = x
                        j += 1
                        x += 1
                    i,j = i+1,j-1
                    direction = Dir.Down
                    right -= 1
                case Dir.Down:
                    while i < down:
                        res[i][j] = x
                        i += 1
                        x += 1
                    i,j = i-1,j-1
                    direction = Dir.Left
                    down -= 1
                case Dir.Left:
                    while j > left:
                        res[i][j] = x
                        j -= 1
                        x += 1
                    i,j = i-1,j+1
                    direction = Dir.Up
                    left += 1
                case Dir.Up:
                    while i > up:
                        res[i][j] = x
                        i -= 1
                        x += 1
                    i,j = i+1,j+1
                    up += 1
                    direction = Dir.Right
        return res                       

File: 59. Spiral matrix II/solution.rs
enum Dir{
    Up,
    Down,
    Left,
    Right,
}
impl Solution {
    pub fn generate_matrix(n: i32) -> Vec<Vec<i32>> {
        let mut res: Vec<Vec<i32>> = vec![vec![0;n as usize] ; n as usize];
        let mut up:usize = 0;
        let mut down:usize = n as usize;
        let mut right:usize = n as usize;
        let mut left:i32 = -1;
        let mut dir:Dir = Dir::Right;
        let mut i:usize = 0;
        let mut j:usize = 0;
        let mut x = 1;
        while x <= n*n{
            match dir{
                Dir::Right =>{
                    while j < right{
                        res[i][j] = x;
                        x += 1;
                        j += 1;
                    }
                    i += 1;
                    j -= 1;
                    right -= 1;
                    dir = Dir::Down;
                }
                Dir::Down =>{
                    while i < down{
                        res[i][j] = x;
                        x += 1;
                        i += 1;
                    }
                    i -= 1;
                    j -= 1;
                    down -= 1;
                    dir = Dir::Left;
                }
                Dir::Left =>{
                    while j as i32 > left{
                        res[i][j] = x;
                        x += 1;
                        j -= 1;
                    }
                    i -= 1;
                    j += 1;
                    left += 1;
                    dir = Dir::Up;
                }
                Dir::Up =>{
                    while i > up{
                        res[i][j] = x;
                        x += 1;
                        i -= 1;
                    }
                    i += 1;
                    j += 1;
                    up += 1;
                    dir = Dir::Right;
                }
            } 
        }res   
        
    }
}